src/compressed_tensors/utils/offload.py

```diff
@@ -49,68 +103,252 @@ def get_execution_device(module: Module) -> torch.device:
     """
-    :param module: layer to check
-    :return: device layer is offloaded to onto after forward pass
+    :param module: module to check
+    :return: device module is offloaded to onto after forward pass
     """
-    if is_module_offloaded(module):
+    if has_offloaded_params(module):
         first_key = list(module._hf_hook.weights_map.keys())[0]
         prefix_dataset = module._hf_hook.weights_map.dataset
         return prefix_dataset[first_key].device
     return next(module.parameters()).device
 
 
-def update_prefix_dict(module: Module, key: str, data: torch.Tensor):
+@check_accelerate(fallback=None)
+def update_prefix_dict(module: torch.nn.Module, key: str, data: torch.Tensor):
     """
     Updates the offloaded state dict for a given module. Parameter named key is replaced
     by data. This is neccesary because parameter updates for offloaded modules do not
     persist automatically between loads. This function only affects the offloaded
     state dict and not the current state of the loaded module.
 
-    :param module: layer containing the parameter to update
+    :param module: module containing the parameter to update
     :param key: name of parameter to update
     :param data: tensor to update parameter with in the offloaded state dict
     """
-    if not is_module_offloaded(module):
+    if not has_offloaded_params(module):
         raise ValueError("Prefix dict is only applicable to offloaded modules")
     prefix_dict = module._hf_hook.weights_map
     prefix_dict.dataset[f"{prefix_dict.prefix}{key}"] = data
 
 
 def update_parameter_data(
-    module: Module, new_param_data: torch.Tensor, param_name: str
+    module: torch.nn.Module, new_param_data: torch.Tensor, param_name: str
 ):
     """
-    Updates the paramter value named param_name for a given module. This function
-    updates both the current loaded module state and the offloaded state dict if
-    the module is offloaded. This is neccesary because parameter updates for offloaded
-    modules do not persist automatically between loads.
+    Update the data of an existing parameter and its offload dict. Supports both
+    parameters of offloaded modules and non-offloaded modules
 
-    :param module: layer containing the parameter to update
+    :param module: module containing the parameter to update
     :param new_param_data: tensor to update parameter with
-    :param param_name: name of layer parameter to update
+    :param param_name: name of module parameter to update
     """
-    if not hasattr(module, param_name):
-        return
+    update_offload_parameter(module, param_name, new_param_data)
 
-    device = next(module.parameters()).device
 
-    offloaded = False
-    if is_module_offloaded(module):
-        offload_device = get_offloaded_device(module)
-        offloaded = True
+""" Candidates for Upstreaming """
+
+
+def register_offload_parameter(
+    module: torch.nn.Module,
+    name: str,
+    parameter: torch.nn.Parameter,
+):
```

Comment:
When registering the parameters during initialization, don't we know the device already, depending on if the module has been offloaded or not? 

We can't pass that device to `update_offload_parameter` to be used when updating the weights_map?

Reply:
I've just now rewritten these parts to a bit clearer.

> don't we know the device already, depending on if the module has been offloaded or not?

During initialization, the `_initialize_scale_zero_point` function determines the *onload device*
```python3
# begin on the same device as other parameters or cpu if offloaded.
# in the offloaded case, there's no point moving tensors to the execution device
# if they're going to be immediately offloaded by `register_offload_parameter`
params_device = next(module.parameters()).device
device = "cpu" if has_offloaded_params(module) else params_device
```

It's the job of `register_offload_parameter` (and by extension `update_offload_parameter`, `offload_to_weights_map`) to determine the *offload* device.
```python3
if isinstance(weights_map, dict):
  if key in weights_map:
      offload_device = weights_map[key].device
  else:
      tens = next(iter(weights_map.values()), None)
      offload_device = tens.device if tens is not None else default_device
```